### ISSUE TITLE
Layer norm weights not excluded from weight decay

### ISSUE THREAD

**Issue by michaellavelle** [karpathy/minGPT#23]
Currently weight decay is being applied to the LayerNorm weights.

I've raised https://github.com/karpathy/minGPT/pull/18 and https://github.com/karpathy/minGPT/pull/16 to resolve this issue.

**Comment by michaellavelle** [comment]
I did a bit of digging into the history - looks like optimization_tf.py was created by @jplu in the "Create a NER example..." commit list here:

https://github.com/jplu/transformers/commits/6be7cdda66f3f4bd3ba4073274bf73be0843c5f9?after=6be7cdda66f3f4bd3ba4073274bf73be0843c5f9+419&branch=6be7cdda66f3f4bd3ba4073274bf73be0843c5f9

With this specific commit:

https://github.com/jplu/transformers/commit/ecb923da9cea390742a1262327a139852c5493e9

This was then merged into the main HuggingFace repo here:

https://github.com/huggingface/transformers/pull/2046/files

I'm assuming that for these TF versions there isn't an issue excluding by parameter name in this way - it's just that the PyTorch parameter names are a different format.    Looks like this automatic exclusion of weight decay is not in the corresponding optimization.py PyTorch file.

**Comment by michaellavelle** [comment]
With regard to excluding positional/token embeddings,  it's a bit tricky.   At first I thought we could just add torch.nn.modules.sparse.Embedding to the no_decay_module_types defined in PR #18  - but as the positional embedding is defined as a standalone Parameter instead of a module, this wouldn't work.

We could potentially exclude all standalone Parameters from weight decay by default ( perhaps with an explicit override list of parameter names we want to include) ?

Or we could invert the problem and specify a list of module types to be included and parameter_names to be excluded in weight decay.

Viewing the problem this way,  for the current setup, we would only include Parameters with Linear module types, and exclude any with the name including "bias".   I just played around with this and can raise a PR for this setup if you'd like to go this direction?

Although that resolves the issue for the current setup,  it's not really a viable solution longer term as the list of modules to include wouldn't be restricted to just Linear.

We need to find some common criteria for modules to be included/excluded - or maybe force this configuration to be part of the trainer config.

**Comment by karpathy** [comment]
Yes the "naked" pos_emb is a bit tricky. I have a specific not amazing but okay solution in mind that I'll push soon as a quick fix.

**Comment by michaellavelle** [comment]
Ok, great.  In case you wanted to go with the "exclude naked Parameters unless overridden solution",  here's what I had working locally:

https://github.com/karpathy/minGPT/pull/24

This excludes LayerNorm, Embedding,  naked Parameters and bias by default,  with a list of parameters to be explicitly included configurable on the TrainerConfig

**Comment by karpathy** [comment]
Ok here is my solutions. Basically I couldn't find a good generic way to make model-free assumptions about what params should or should not be decayed. In that case it makes most sense to couple this decision inside the model code itself. And if we're doing that we might as well look into making things incrementally more compatible with Lightning, to which we may want to switch later. 

https://github.com/karpathy/minGPT/commit/bbbdac74fa9b2e55574d70056163ffbae42310c1

**Comment by michaellavelle** [comment]
Great - glad you found a way round this.  I'm sure many people will be referencing this training code, so having the weight decay logic decoupled from the model when it's very model specific makes sense.

Thanks for sorting this, and again for the really useful project.

btw, I created https://github.com/karpathy/minGPT/pull/25 - hopefully not too contentious.

**Comment by Rachnas** [comment]
> Yes the "naked" pos_emb is a bit tricky. I have a specific not amazing but okay solution in mind that I'll push soon as a quick fix.

I did a small experiment, changed pos_emb from nn.Parameter to nn.Embedding and other changes accordingly. Got 100% on play_math test set also.
